File: src/main/java/serendustry/machine/MetaTileEntityTranscendentPlasmaMixer.java
package serendustry.machine;

import javax.annotation.Nonnull;

import net.minecraft.block.state.IBlockState;
import net.minecraft.util.ResourceLocation;

import gregtech.api.metatileentity.MetaTileEntity;
import gregtech.api.metatileentity.interfaces.IGregTechTileEntity;
import gregtech.api.metatileentity.multiblock.IMultiblockPart;
import gregtech.api.metatileentity.multiblock.RecipeMapMultiblockController;
import gregtech.api.pattern.BlockPattern;
import gregtech.api.pattern.FactoryBlockPattern;
import gregtech.api.pattern.PatternMatchContext;
import gregtech.client.renderer.ICubeRenderer;
import gregtech.client.renderer.texture.Textures;
import gregtech.common.blocks.BlockMachineCasing;
import gregtech.common.blocks.MetaBlocks;

public class MetaTileEntityTranscendentPlasmaMixer extends RecipeMapMultiblockController {

    public MetaTileEntityTranscendentPlasmaMixer(ResourceLocation metaTileEntityId) {
        super(metaTileEntityId, SerendustryRecipeMaps.PLASMA_MIXER_RECIPES);
    }

    @Override
    public MetaTileEntity createMetaTileEntity(IGregTechTileEntity iGregTechTileEntity) {
        return new MetaTileEntityTranscendentPlasmaMixer(metaTileEntityId);
    }

    @Nonnull
    @Override
    protected BlockPattern createStructurePattern() {
        return FactoryBlockPattern.start()
                .aisle("XXX", "XXX", "XXX")
                .aisle("XXX", "XXX", "XXX")
                .aisle("XXX", "XXX", "XXX")
                .aisle("XXX", "XXX", "XXX")
                .aisle("XXX", "XSX", "XXX")
                .where('S', selfPredicate())
                .where('X', states(getCasingState()).setMinGlobalLimited(60).or(autoAbilities()))
                .build();
    }

    @Override
    public ICubeRenderer getBaseTexture(IMultiblockPart iMultiblockPart) {
        return Textures.INERT_PTFE_CASING; // todo
    }

    protected IBlockState getCasingState() {
        // todo: add custom Physically Improbable Noble Gases Casing
        return MetaBlocks.MACHINE_CASING.getState(BlockMachineCasing.MachineCasingType.UIV);
    }

    @Override
    public void invalidateStructure() {
        super.invalidateStructure();
    }

    @Override
    protected void formStructure(PatternMatchContext context) {
        super.formStructure(context);
    }
}

File: src/main/kotlin/serendustry/machine/MetaTileEntityTranscendentPlasmaMixer.kt
package serendustry.machine

import gregtech.api.metatileentity.interfaces.IGregTechTileEntity
import gregtech.api.metatileentity.multiblock.IMultiblockPart
import gregtech.api.metatileentity.multiblock.RecipeMapMultiblockController
import gregtech.api.pattern.BlockPattern
import gregtech.api.pattern.FactoryBlockPattern
import gregtech.client.renderer.ICubeRenderer
import gregtech.client.renderer.texture.Textures
import gregtech.common.blocks.BlockMachineCasing
import gregtech.common.blocks.MetaBlocks
import net.minecraft.block.state.IBlockState
import net.minecraft.util.ResourceLocation

class MetaTileEntityTranscendentPlasmaMixer(rl: ResourceLocation): RecipeMapMultiblockController(rl, PLASMA_MIXER_RECIPES) {

    override fun createMetaTileEntity(te: IGregTechTileEntity) = MetaTileEntityTranscendentPlasmaMixer(metaTileEntityId)

    override fun createStructurePattern(): BlockPattern =
        FactoryBlockPattern.start()
            .aisle("XXX", "XXX", "XXX")
            .aisle("XXX", "XXX", "XXX")
            .aisle("XXX", "XXX", "XXX")
            .aisle("XXX", "XXX", "XXX")
            .aisle("XXX", "XSX", "XXX")
            .where('S', selfPredicate())
            .where('X', states(getCasingState()).setMinGlobalLimited(60).or(autoAbilities()))
            .build()

    override fun getBaseTexture(part: IMultiblockPart?): ICubeRenderer = Textures.INERT_PTFE_CASING // todo

    // todo: add custom Physically Improbable Noble Gases Casing
    private fun getCasingState(): IBlockState = MetaBlocks.MACHINE_CASING.getState(BlockMachineCasing.MachineCasingType.UIV)
}
